src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,972 @@
+                            HIDDEN,
+                            SIGMOID,  // TODO: make it random??
+                            (fromNeuron.Depth() + toNeuron.Depth()) / 2,
+                            0,
+                            0);
+      genome.AddHiddenNeuron(neuronGene);
+
+      LinkGene inputLink(genome.aLinkGenes[linkIdx].FromNeuronId(),
+                         aNeuronInnovations[innovIdx].newNeuronId,
+                         aNeuronInnovations[innovIdx].newInputLinkInnovId,
+                         1,
+                         true);
+      LinkGene outputLink(aNeuronInnovations[innovIdx].newNeuronId,
+                          genome.aLinkGenes[linkIdx].ToNeuronId(),
+                          aNeuronInnovations[innovIdx].newOutputLinkInnovId,
+                          genome.aLinkGenes[linkIdx].Weight(),
+                          true);
+      genome.AddLink(inputLink);
+      genome.AddLink(outputLink);
+      return;
+    }
+
+    // If new innovation, create.
+    NeuronInnovation neuronInnov = AddNeuronInnovation(splitLinkInnovId);
+
+    LinkInnovation inputLinkInnov;
+    inputLinkInnov.fromNeuronId = genome.aLinkGenes[linkIdx].FromNeuronId();
+    inputLinkInnov.toNeuronId = neuronInnov.newNeuronId;
+    inputLinkInnov.newLinkInnovId = neuronInnov.newInputLinkInnovId;
+    aLinkInnovations.push_back(inputLinkInnov);
+
+    LinkInnovation outputLinkInnov;
+    outputLinkInnov.fromNeuronId = neuronInnov.newNeuronId;
+    outputLinkInnov.toNeuronId = genome.aLinkGenes[linkIdx].ToNeuronId();
+    outputLinkInnov.newLinkInnovId = neuronInnov.newOutputLinkInnovId;
+    aLinkInnovations.push_back(outputLinkInnov);
+
+    NeuronGene neuronGene(neuronInnov.newNeuronId,
+                          HIDDEN,
+                          SIGMOID,  // TODO: make it random??
+                          (fromNeuron.Depth() + toNeuron.Depth()) / 2,
+                          0,
+                          0);
+    genome.AddHiddenNeuron(neuronGene);
+
+    LinkGene inputLink(genome.aLinkGenes[linkIdx].FromNeuronId(),
+                       neuronInnov.newNeuronId,
+                       neuronInnov.newInputLinkInnovId,
+                       1,
+                       true);
+    LinkGene outputLink(neuronInnov.newNeuronId,
+                        genome.aLinkGenes[linkIdx].ToNeuronId(),
+                        neuronInnov.newOutputLinkInnovId,
+                        genome.aLinkGenes[linkIdx].Weight(),
+                        true);
+    genome.AddLink(inputLink);
+    genome.AddLink(outputLink);
+  }
+
+  // Mutate: enable disabled, or disable enabled link.
+  void MutateEnableDisable(Genome& genome, bool enabled, double mutateProb) {
+    std::vector<ssize_t> linkIndexs;
+    for (ssize_t i=0; i<genome.NumLink(); ++i) {
+      if (genome.aLinkGenes[i].Enabled() == enabled) {
+        linkIndexs.push_back(i);
+      }
+    }
+
+    double p = mlpack::math::Random();
+    if (p<mutateProb && linkIndexs.size()>0) {
```

Comment:
I think it worth to check the mutate probability before we generate the `linkIndexs`list.